stub_uploader/metadata.py

```diff
@@ -190,30 +235,27 @@ def verify_external_req(
             f"There is no upstream distribution on PyPI, so cannot verify {req}"
         )
 
-    resp = requests.get(f"https://pypi.org/pypi/{upstream_distribution}/json")
-    if resp.status_code != 200:
+    if req.name not in EXTERNAL_REQ_ALLOWLIST and not _unsafe_ignore_allowlist:
         raise InvalidRequires(
-            f"Expected dependency {req} to be accessible on PyPI, but got {resp.status_code}"
+            f"Expected dependency {req} to be present in the allowlist"
         )
 
+    resp = requests.get(f"https://pypi.org/pypi/{upstream_distribution}/json")
+    validate_response(resp, req)
     data = resp.json()
 
-    # TODO: consider allowing external dependencies for stubs for packages that do not ship wheels.
-    # Note that we can't build packages from sdists, since that can execute arbitrary code.
-    # We could do some hacky setup.py parsing though...
     # TODO: PyPI doesn't seem to have version specific requires_dist. This does mean we can be
     # broken by new releases of upstream packages, even if they do not match the version spec we
     # have for the upstream distribution.
-    if req.name not in [
+    if req.name in [
         Requirement(r).name for r in (data["info"].get("requires_dist") or [])
     ]:
-        raise InvalidRequires(
-            f"Expected dependency {req} to be listed in {upstream_distribution}'s requires_dist"
-        )
+        return  # Ok!
 
-    if req.name not in EXTERNAL_REQ_ALLOWLIST and not _unsafe_ignore_allowlist:
+    if req_canonical_name not in get_sdist_requires_canonical(data["urls"][-1], req):
```

Comment:
> `data["urls"][-1]`

Two assumptions here:
1. The Source Distribution is always last, Built Distributions (wheels) will come first
2. If there's no Source Distribution (like `pywin32`, this will incidentally take a wheel. But if wheels are available we shouldn't get here in the first place.

I think it's worth checking for `python_version == "source"` to be more rigorous.